cmd/query/app/server.go

```diff
@@ -349,11 +352,22 @@ func (s *Server) Close() error {
 	var errs []error
 	errs = append(errs, s.queryOptions.TLSGRPC.Close())
 	errs = append(errs, s.queryOptions.TLSHTTP.Close())
+
+	// Stop gRPC server
+	s.logger.Info("Stopping gRPC server")
 	s.grpcServer.Stop()
+
+	// Log and close HTTP server
+	s.logger.Info("Closing HTTP server")
 	errs = append(errs, s.httpServer.Close())
```

Comment:
if feels to me that here is the actual issue. When we're running on a single port with cmux, we create a single Listener. But the way at least HTTP server works is when you pass it a listener it assumes it owns it and will close it when stopping the server. Since we have several servers they probably all try to close the listener, but in random order due to goroutine scheduling. I suspect if we add sleeps in this function after each call to Close (and maybe change the order of closing servers) then we might be able to reproduce the error in the failed test. 

If my assumption is correct, I still don't know how to actually fix the issue, since we don't control how servers deal with Listener. One thing we could do is to explicitly check for "conn already closed" error and not return it (but only when running on a single port).